Публикация

delannushka opened the pull request:
---
:mortar_board: [Публикация](https://up.htmlacademy.ru/yii/4/user/2069609/tasks/15)

delannushka committed:
Выполяет 7.6. Публикация

nkolosov commented on models/forms/TaskCreateForm.php:
```diff
@@ -0,0 +1,76 @@
+        //$task->files = $this->files;
+        $task->budget = $this->budget;
+        $task->user_id = Yii::$app->user->id;
+        $task->save();
```
1. Нужно проверку, что сохранение прошло успешно.
2. Загрузка файлов и создание задания должны быть обернуты в одну транзакцию БД

nkolosov commented on models/forms/TaskCreateForm.php:
```diff
@@ -0,0 +1,76 @@
+        ];
+    }
+
+        public function rules()
```
Поехало форматирование немного

delannushka committed:
Вносит правки 7.6. Публикация

nkolosov commented on controllers/TasksController.php:
```diff
@@ -48,4 +74,30 @@ public function actionView($id){
+    public function actionCreate()
+    {
+        $taskCreateForm = new TaskCreateForm();
+        if (Yii::$app->request->getIsPost()) {
```
А как попасть на собственно форму создания? Тут if неправильно обернут, в случае обычного запроса (перехода по ссылке) ничего не работает

nkolosov commented on controllers/TasksController.php:
```diff
@@ -48,4 +74,30 @@ public function actionView($id){
+        $taskCreateForm = new TaskCreateForm();
+        if (Yii::$app->request->getIsPost()) {
+            $taskCreateForm->load(Yii::$app->request->post());
+            $taskCreateForm->files = UploadedFile::getInstances($taskCreateForm, 'files');
```
Лучше всё перенести в модель, включая транзакцию. Метод в модели пусть возвращает true / false

А в контроллере только обработку ошибки сделать

nkolosov commented on models/forms/TaskCreateForm.php:
```diff
@@ -0,0 +1,83 @@
+        $task->budget = $this->budget;
+        $task->user_id = Yii::$app->user->id;
+        $task->save();
+        if ($task->save()) {
```
Выше уже вызвали save, не надо делать это 2 раза, нужно сразу проверить результат

nkolosov commented on models/forms/TaskCreateForm.php:
```diff
@@ -0,0 +1,83 @@
+        if ($task->save()) {
+            return $task;
+        } else {
+            throw new ServerErrorHttpException('Loading error');
```
Http - Это уровень уже контроллера, модель должна возвращать какой-нибудь внутренний (самописный) exception

nkolosov commented on models/forms/TaskCreateForm.php:
```diff
@@ -0,0 +1,83 @@
+                $fileRecord = new File();
+                $fileRecord->task_id = $taskId;
+                $fileRecord->title = '@webroot/uploads/' . $newName;
+                $fileRecord->save();
```
Нужна обработка ошибки

delannushka committed:
Вносит правки 7.6. Публикация

nkolosov commented on controllers/TasksController.php:
```diff
@@ -2,13 +2,39 @@
+                    [   'allow' => true,
+                        'actions' => ['create'],
+                        'matchCallback' => function () {
+                            return Yii::$app->user->identity->role == User::ROLE_CUSTOMER;
```
Если напрямую попытаться зайти на /tasks/create получаем ошибку (если не залогиненны)

Attempt to read property "role" on null

А должен происходить редирект на логин

nkolosov commented on controllers/TasksController.php:
```diff
@@ -48,4 +74,25 @@ public function actionView($id){
+                try {
+                    $taskCreateForm->doTransaction($taskCreateForm);
+                } catch (Exception $e) {
+                    return throw new Exception('Loading error');
```
return throw new - Так не делается.

Тут нужно как раз редирект сделать на страницу с ошибкой (или сделать throw new InternalServerHttpException

nkolosov commented on models/forms/TaskCreateForm.php:
```diff
@@ -0,0 +1,105 @@
+        if ($task->save()) {
+            return $task;
+        } else {
+            throw new Exception('Task loading error');
```
Не loading, а saving

nkolosov commented on controllers/TasksController.php:
```diff
@@ -2,13 +2,39 @@
 {
+    public function behaviors()
+    {
+        return [
```
Кроме того, рекомендую делать array_merge(parent::behaviors(), [...]) чтобы не потерять значения из класса-родителя

delannushka committed:
Вносит правки 7.6. Публикация.
Все равно пришлось вставить строчку 'roles' => ['@'], в behaviors для действия create. Иначе ошибка при попытке входа незалогиненного пользователя на страницу tasks/create
